### ISSUE TITLE
scorch in-memory mode

### ISSUE THREAD

**Issue by steveyen** [blevesearch/bleve#1165]
Would be interesting to have an in-memory-only mode for scorch.

Some notes...

- there's some baton handoffs between the introducer / persister / file-merger that need careful consideration.  e.g., can't just naively turn off the persister, because the introducer might end up blocking trying to pass a baton to a non-existent persister.

- scorch has a file-based segment merger, but doesn't have a separate in-memory merger goroutine

- the persister goroutine does perform some in-memory merging -- perhaps it would be a natural place to "keep the persister -- but it doesn't actually do any persistence"

Also, perhaps related (or might need to be a separate enhancement/issue)... at least one app (couchbase server) has a need for a fast "reset" of an in-memory-only index back to empty state.  See discussion at... https://github.com/blevesearch/bleve/pull/1163

**Comment by ellipticasec** [comment]
Hi, I would most definitely have a use case for this. Where I need to create very small, short-lived indexes and would like to avoid the penalty of writing to disk, etc... I actually use a RAM Disk already, but looking for some improvements.

**Comment by fproulx** [comment]
Love it.

**Comment by abourget** [comment]
yeah, way to go :)

**Comment by fproulx-dfuse** [comment]
Bump @steveyen is this on the roadmap ?

**Comment by orius123** [comment]
@mschoch / @steveyen are there any plans of adding in the near future?

**Comment by mschoch** [comment]
@orius123 Yes, my hope is that we get this feature in the next 6 months.  It is already working on Bluge, but we have to backport many other changes to bring it into bleve.  Bleve/Scorch/Zap today all directly work with the filesystem, in Bluge we introduce a `Directory` abstraction.  By making this change, an in-memory only index doesn't require any other substantial changes to the index/search code, the Directory simply omits writing to disk.  We do however expect there would need to be more work spent to tune the in-memory index to work well, as many other scorch settings (merge aggressiveness, etc) may default  to values which are not appropriate for in-memory only use.

Unfortunately, this feature is not a high-priority for the Couchbase use-case, which has made it hard for me to spend much time on this.  If a high-enough level sponsor wanted this, or there was some contract work lined up for it, this could be delivered sooner and with a more concrete timeline.
